panic: Layers by definition are suffixes of names

I ran into this message while doing a bunch of DNS lookups. I'm not sure what causes this error, but it would be helpful if zmap just returned an error JSON for the corresponding lookup, rather than dying with a panic.

```
panic: Layers by definition are suffixes of names

goroutine 704 [running]:
github.com/zmap/zdns/modules/miekg.nextAuthority(0xc0017fccb0, 0xb, 0xc0006e9760, 0xb, 0xc001e4dd8f, 0x2)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:788 +0x2ed
github.com/zmap/zdns/modules/miekg.(*Lookup).cachedRetryingLookup(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0xc001e4dd80, 0x11, 0xc0006e9760, 0xb, 0x4, 0x0, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:735 +0x631
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0xc001e4dd80, 0x11, 0x4, 0xc0006e9760, 0xb, 0x944f80, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:964 +0x2ac
github.com/zmap/zdns/modules/miekg.(*Lookup).iterateOnAuthorities(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0x3, 0x944f80, 0x0, 0x0, 0x944f80, 0x0, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:933 +0xbe0
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0xc001d56900, 0x10, 0x3, 0xc001d56770, 0x9, 0x944f80, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:998 +0xaf4
github.com/zmap/zdns/modules/miekg.(*Lookup).iterateOnAuthorities(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0x2, 0x944f80, 0x0, 0x0, 0x944f80, 0x0, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:933 +0xbe0
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0xc0032d0840, 0x10, 0x2, 0xc000ab58a0, 0x3, 0x944f80, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:998 +0xaf4
github.com/zmap/zdns/modules/miekg.(*Lookup).iterateOnAuthorities(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0x1, 0x944f80, 0x0, 0x0, 0x944f80, 0x0, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:933 +0xbe0
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc00286af00, 0x10001, 0xc0017fccb0, 0xb, 0x6fb40b, 0x10, 0x1, 0x6f860a, 0x1, 0x944f80, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:998 +0xaf4
github.com/zmap/zdns/modules/miekg.(*Lookup).DoMiekgLookup(0xc00286af00, 0xc0017fccb0, 0xb, 0xc0000ac870, 0x6f8681, 0x2, 0x945cc0, 0x656c00, 0xc0000c41e0, 0x10000006fb40b, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:1009 +0x285
github.com/zmap/zdns/modules/miekg.(*Lookup).DoLookup(0xc00286af00, 0xc0017fccb0, 0xb, 0x0, 0x0, 0x2e3, 0x0, 0x0, 0x6f90b6, 0x7, ...)
        /home/nikita/go/src/github.com/zmap/zdns/modules/miekg/miekg.go:1091 +0x46
github.com/zmap/zdns.doLookup(0xc0000aa810, 0xc0000d0360, 0xc0000a00c0, 0xc0000a0120, 0xc0000c4960, 0xc0000be2d0, 0x2ab, 0x7fee1f220888, 0xc00006aba8)
        /home/nikita/go/src/github.com/zmap/zdns/lookup.go:114 +0x832
created by github.com/zmap/zdns.DoLookups
        /home/nikita/go/src/github.com/zmap/zdns/lookup.go:176 +0x387```

--- a/modules/miekg/miekg.go
+++ b/modules/miekg/miekg.go
@@ -875,7 +875,7 @@ func (s *Lookup) cachedRetryingLookup(dnsType uint16, dnsClass uint16, name stri
 	if err != nil {
 		s.VerboseLog(depth+2, err)
 		var r Result
-		return r, isCached, zdns.STATUS_AUTHFAIL, nil
+		return r, isCached, zdns.STATUS_AUTHFAIL, err
 	}
 	if name != layer && authName != layer {
 		if authName == "" {

--- a/modules/miekg/miekg.go
+++ b/modules/miekg/miekg.go
@@ -871,7 +871,12 @@ func (s *Lookup) cachedRetryingLookup(dnsType uint16, dnsClass uint16, name stri
 	// Now, we check the authoritative:
 	name = strings.ToLower(name)
 	layer = strings.ToLower(layer)
-	authName := nextAuthority(name, layer)
+	authName, err := nextAuthority(name, layer)
+	if err != nil {
+		s.VerboseLog(depth+2, err)
+		var r Result
+		return r, isCached, zdns.STATUS_AUTHFAIL, nil
+	}
 	if name != layer && authName != layer {
 		if authName == "" {
 			s.VerboseLog(depth+2, "Can't parse name to authority properly. name: ", name, ", layer: ", layer)
@@ -908,35 +913,35 @@ func nameIsBeneath(name string, layer string) (bool, string) {
 	return false, ""
 }
 
-func nextAuthority(name string, layer string) string {
+func nextAuthority(name string, layer string) (string, error) {
 	// We are our own authority for PTRs
 	// (This is dealt with elsewhere)
 	if strings.HasSuffix(layer, "in-addr.arpa") {
-		return name
+		return name, nil
 	}
 
 	idx := strings.LastIndex(name, ".")
 	if idx < 0 || (idx+1) >= len(name) {
-		return name
+		return name, nil
 	}
 	if layer == "." {
-		return name[idx+1:]
+		return name[idx+1:], nil
 	}
 
 	if !strings.HasSuffix(name, layer) {
-		panic("Layers by definition are suffixes of names")
+		return "", errors.New("Server did not provide appropriate resolvers to continue recursion")
 	}
 
 	// Limit the search space to the prefix of the string that isnt layer
 	idx = strings.LastIndex(name, layer) - 1
 	if idx < 0 || (idx+1) >= len(name) {
 		// Out of bounds. We are our own authority
-		return name
+		return name, nil
 	}
 	// Find the next step in the layer
 	idx = strings.LastIndex(name[0:idx], ".")
 	next := name[idx+1:]
-	return next
+	return next, nil
 }
 
 func (s *Lookup) checkGlue(server string, depth int, result Result) (Result, zdns.Status) {

